openproblems/tasks/spatial_decomposition/datasets/_sc_to_sp_utils.py

```diff
@@ -0,0 +1,169 @@
+        - `adata_spatial.uns["sc_reference"]`: original sc adata for reference.
+        - `adata_spatial.obsm["proportions_true"]`: true proportion values.
+        - `adata_spatial.obsm["n_cells"]`: number of cells from each type at
+           every location
+        - `adata_spatial.obs["proportions_true"]`:
+           total number of cells at each location
+
+    The cell type labels are stored in adata_sc.obs["label"].
+    """
+
+    # set random generator seed
+    rng = np.random.default_rng(seed)
+
+    # get single cell expression data
+    X = adata.X
+    if issparse(X):
+        X = X.toarray()
+    # get cell annotations/labels
+    labels = adata.obs[type_column].values
+    # get unique labels
+    uni_labs = np.unique(labels)
+    # count number of labels
+    n_labs = len(uni_labs)
+    # get number of genes
+    n_genes = adata.shape[1]
+
+    # create dict with indices of each label
+    label_indices = dict()
+    for label in uni_labs:
+        label_indices[label] = np.where(labels == label)[0]
+
+    # adjust alpha to vector if single scalar
+    if not hasattr(alpha, "__len__"):
+        alpha = np.ones(n_labs) * alpha
+    else:
+        assert len(alpha) == n_labs, "alpha must be same size as number of cell types"
+
+    # generate probability of sampling label at each spot
+    sp_props = rng.dirichlet(alpha, size=n_obs)
+    # number of cells present at each spot
+    n_cells = rng.integers(cell_lb, cell_ub, size=n_obs)
+
+    # initialize spatial expression matrix
+    sp_x = np.zeros((n_obs, n_genes))
+    # initialize spatial proportion matrix
+    sp_p = np.zeros((n_obs, n_labs))
+    # initialize spatial cell number matrix
+    sp_c = np.zeros(sp_p.shape)
+
+    # generate expression vector for each spot (s)
+    for s in range(n_obs):
+        # number of cells from each label at s
+        raw_s = rng.multinomial(n_cells[s], pvals=sp_props[s, :])
+        # store number of cells from each type at s
+        sp_c[s, :] = raw_s
+        # compute proportion of each type at s
+        prop_s = raw_s / n_cells[s]
+        # store proportion of each type at s
+        sp_p[s, :] = prop_s
+
+        # initialize transcript pool at s
+        pool_s = np.zeros(n_genes)
+
+        # add molecules to transcript pool
+        for lab, n in enumerate(raw_s):
+            # get indices of cells from which transcripts should be added
+            idx_sl = rng.choice(label_indices[uni_labs[lab]], size=n)
+            # add molecules to pool
+            pool_s += X[idx_sl, :].sum(axis=0)
+
+        # number of UMIs at spot s
+        n_umis = rng.integers(umi_lb, umi_ub)
+        # compute probability of sampling UMI from gene
+        prob_pool_s = pool_s / pool_s.sum()
+
+        # sample transcripts from pool
+        sp_x[s, :] = np.random.multinomial(n=n_umis, pvals=prob_pool_s)
+
+    obs_names = ["spatial_{}".format(x) for x in range(n_obs)]
+    adata_spatial = ad.AnnData(
+        sp_x,
```

Comment:
```suggestion
        csr_matrix(sp_x),
```

Reply:
can the count be sparse?

Reply:
actually, I initially had the matrix as sparse, but wasn't sure all the methods could handle it. Hence why I kept it as dense, but if you think it'll be fine that's great!